### PR TITLE
feature errorsSpace, warningsSpace

### LINKED ISSUE #14845: more user friendly error on execute module fail
## Feature request

right now  webpack puts generated code when executing module in error https://github.com/webpack/webpack/blob/main/lib/javascript/JavascriptModulesPlugin.js#L439

this stack trace will be printed in stdout

**What is the expected behavior?**

more user friendly output

**What is motivation or use case for adding/changing the behavior?**

improve DX

**How should this be implemented in your opinion?**

with current design I suggest to limit stack trace string length in stats

**Are you willing to work on this yourself?**
yes

### PR DESCRIPTION
**Pull request by vankop**
**What kind of change does this PR introduce?**
feature
closes #14845 

**Did you add tests for your changes?**
yes

**Does this PR introduce a breaking change?**
no

**What needs to be documented once your changes are merged?**
`stats.errorsSpace` , `stats.warningsSpace` Space to display errors/warnings in number of lines

### CODE REVIEW

File: lib/stats/DefaultStatsFactoryPlugin.js [hunk 1]
```diff
@@ -1779,6 +1781,46 @@ const spaceLimited = (
...
 
+const errorsSpaceLimit = (errors, max) => {
+	if (!Number.isFinite(max)) return errors;
+	// Can not fit into limit
+	// print only messages
+	if (errors.length > max)
+		return errors.map(error => {
+			if (error.details === undefined) return error;
+			const { details, ...rest } = error;
+			return { details: "", ...rest };
```
Comment (sokra) [inline lib/stats/DefaultStatsFactoryPlugin.js:1792]:
```suggestion
			return { ...error, details: "" };
```

File: lib/stats/DefaultStatsFactoryPlugin.js [hunk 2]
```diff
@@ -1779,6 +1781,46 @@ const spaceLimited = (
...
+			const splitted = errors[i].details.split("\n");
+			const len = splitted.length;
+			fullLength += len;
+			if (fullLength > max) {
+				result = errors.slice(0, i);
+				const { details, ...rest } = errors[i++];
+				result.push({
+					details: splitted.slice(0, len - (fullLength - max)).join("\n"),
+					...rest
+				});
```
Comment (sokra) [inline lib/stats/DefaultStatsFactoryPlugin.js:1810]:
```suggestion
				result.push({
				  ...errors[i++],
					details: splitted.slice(0, len - (fullLength - max)).join("\n")
				});
```

No need to create a temporary `rest` object.

File: lib/stats/DefaultStatsFactoryPlugin.js [hunk 3]
```diff
@@ -1779,6 +1781,46 @@ const spaceLimited = (
...
+				for (; i < errors.length; i++) {
+					const { details, ...rest } = errors[i];
+					if (details === undefined) result.push(errors[i]);
+					result.push({ details: "", ...rest });
```
Comment (sokra) [inline lib/stats/DefaultStatsFactoryPlugin.js:1814]:
```suggestion
					const { details } = errors[i];
					if (details === undefined) result.push(errors[i]);
					result.push({ ...errors[i], details: "" });
```

**Review by sokra — CHANGES_REQUESTED** [review]

File: lib/stats/DefaultStatsFactoryPlugin.js [hunk 4]
```diff
@@ -1779,6 +1781,46 @@ const spaceLimited = (
...
+	for (; i < errors.length; i++) {
+		if (errors[i].details) {
+			const splitted = errors[i].details.split("\n");
+			const len = splitted.length;
+			fullLength += len;
+			if (fullLength > max) {
+				result = errors.slice(0, i);
+				const { details, ...rest } = errors[i++];
+				result.push({
+					details: splitted.slice(0, len - (fullLength - max)).join("\n"),
```
Comment (sokra) [inline lib/stats/DefaultStatsFactoryPlugin.js:1808]: We should add some note that some lines has been omitted of the details.

File: lib/stats/DefaultStatsFactoryPlugin.js [hunk 5]
```diff
@@ -1779,6 +1781,46 @@ const spaceLimited = (
...
+					const { details, ...rest } = errors[i];
+					if (details === undefined) result.push(errors[i]);
+					result.push({ details: "", ...rest });
+				}
+				break;
+			}
+		}
+	}
+
+	return result;
```
Comment (sokra) [inline lib/stats/DefaultStatsFactoryPlugin.js:1821]: We should add some hint that not all details are shown.

We currently show a hint for `filteredWarningDetailsCount`. But it need to show the correct flag to enable it... Either `stats.errorDetails: true` or `stats.errorsSpaces: Infinity` depending on the reason of filtering.

Comment (sokra) [reply lib/stats/DefaultStatsFactoryPlugin.js:1821]: Probably add a separate counter for the filtering due to space.

Comment (vankop) [reply lib/stats/DefaultStatsFactoryPlugin.js]: I decided to use `filteredWarningDetailsCount`, looks like it fits well here.

File: lib/stats/DefaultStatsFactoryPlugin.js [hunk 6]
```diff
@@ -862,16 +888,22 @@ const SIMPLE_EXTRACTORS = {
 			if (errorDetails === "auto") {
 				if (warnings) {
 					const warnings = cachedGetWarnings(compilation);
-					object.filteredWarningDetailsCount = warnings
-						.map(e => typeof e !== "string" && e.details)
-						.filter(Boolean).length;
+					const filtered = object.filteredWarningDetailsCount || 0;
```
Comment (sokra) [inline lib/stats/DefaultStatsFactoryPlugin.js:891]: It doesn't seem right to read from `object`.

Best move the whole `errorsDetails` logic here into `errors`/`warnings` and write `filteredWarningDetailsCount` only there.

File: lib/stats/DefaultStatsFactoryPlugin.js [hunk 7]
```diff
@@ -1779,6 +1811,60 @@ const spaceLimited = (
...
+			if (fullLength > max) {
+				result = i > 0 ? errors.slice(0, i) : [];
+				const overlimit = fullLength - max;
+				result.push({
+					...errors[i++],
+					details: `${splitted
+						.slice(0, len - overlimit)
+						.join("\n")}\n+${overlimit} hidden line${
+						overlimit === 1 ? "" : "s"
+					}`
```
Comment (sokra) [inline lib/stats/DefaultStatsFactoryPlugin.js:1845]: The extra line should be added by the `DefaultStatsPrinterPlugin`.
Only add a `filteredDetails` property to the object instead.

Comment (vankop) [reply lib/stats/DefaultStatsFactoryPlugin.js:1845]: this lines is related to partial details. e.g. case 2 errors.
- first has 3 lines of details
- second 2 lines.
- errorsScope=5
then we should get:
<error #1 message> (1)
<error #1 details line #1> (2)
<error #1 details line #2> (3)
+ 1 hidden line (4)

<error #2 message> (5)

File: cspell.json [hunk 8]
```diff
@@ -161,6 +161,7 @@
     "opencollective",
     "opensource",
     "opuuus",
+    "overlimit",
```
Comment (sokra) [inline cspell.json:164]: I guess the spellchecker wants you to write it as `overLimit` instead of `overlimit` since these are two words.

**Comment by alexander-akait** [comment]
@vankop This too :smile:

**Review by alexander-akait — APPROVED** [review]

**Review by TheLarkInn — APPROVED** [review]

**Pull request merged**